relayer-engine/src/walletMonitor/balances.ts

```diff
@@ -0,0 +1,531 @@
+    (tokenAddress: string | null) =>
+      tokenAddress && tokenAddress !== ethers.constants.AddressZero
+  );
+
+  _logger.debug(`Local addresses for chain ${chainId}: ${retVal.join(', ')}`);
+
+  return retVal;
+}
+
+async function calcLocalAddressesTerra(
+  lcd: LCDClient,
+  tokensToMonitor: TokenMonitoringInfo[],
+  chainId: ChainId,
+  tokenBridgeAddress: string,
+) {
+  const _logger = getScopedLogger(["calcLocalAddressesTerra"], logger());
+
+  const output: string[] = [];
+  for (const token of tokensToMonitor) {
+    if (token.chainId === chainId) {
+      if (token.address.startsWith("terra")) {
+        _logger.debug(`${token.address} is a native token`);
+        output.push(token.address);
+      }
+      continue;
+    }
+
+    let hexAddress;
+    try {
+      hexAddress = tryNativeToHexString(
+        token.address,
+        wormholeChainIdtoChainName(token.chainId),
+      );
+    } catch (error) {
+      _logger.error(`Failed to convert address ${token.address} to hex. Error: ${error}`);
+      continue;
+    }
+
+    let foreignAddress;
+    try {
+      foreignAddress = await getForeignAssetTerra(
+        tokenBridgeAddress,
+        lcd,
+        token.chainId,
+        hexToUint8Array(hexAddress)
+      );
+    } catch (e) {
+      _logger.error(`Failed to get Terra foreign address for token ${token.address}. Error: ${e}`);
+    }
+
+    if (!foreignAddress) continue;
+
+    output.push(foreignAddress);
+  }
+
+  _logger.debug(`Local addresses found for chain ${chainId}: ${output.join(', ')}`);
+  return output;
+}
+
+async function pullEVMTokens(
+  privateKey: string,
+  chainConfig: ChainConfigInfo,
+  tokensToMonitor: TokenMonitoringInfo[],
+  provider: ethers.providers.JsonRpcProvider,
+) {
+  const _logger = getScopedLogger(["pullEVMTokens"], logger());
+
+  if (tokensToMonitor.length === 0) {
+    _logger.verbose(`No tokens to monitor for chain ${chainConfig.chainId}!`);
+    return [];
+  }
+
+  if (!chainConfig.tokenBridgeAddress) {
+    _logger.error(`Wont pull EVM tokens, missing tokenBridgeAddress ${JSON.stringify(chainConfig)}`);
+    return [];
+  }
+
+  const tokenBridgeAddress = chainConfig.tokenBridgeAddress!;
+
+  const localAddresses = await calcLocalAddressesEVM(
+    provider,
+    tokensToMonitor,
+    chainConfig.chainId,
+    tokenBridgeAddress,
+  );
+
+  if (!localAddresses.length) {
+    _logger.debug(`No local addresses found for chain ${chainConfig.chainId}!`);
+    return [];
+  }
+
+  let publicAddress: string;
+  try {
+    publicAddress = await new ethers.Wallet(privateKey).getAddress();
+  } catch (error) {
+    _logger.error("Failed to get public address", error);
+    return [];
+  }
+
+  localAddresses.forEach(async (tokenAddress) => {
+    try {
+      const token = await getEthereumToken(tokenAddress, provider);
+      const [decimals, balanceValue, symbol] = await Promise.all([
+        token.decimals(),
+        token.balanceOf(publicAddress),
+        token.symbol()
```

Comment:
Only `balanceOf` should be called here. `symbol` and `decimals` should be called only once during configuration at startup.